File: examples/callbacks/tests/bindings/test_callbacks.py
import unittest
from callbacks import *

# This is defined in UDL as a "callback". It's not possible to have a Rust
# implementation of a callback, they only exist on the foreign side.
class CallAnswererImpl(CallAnswerer):
    def __init__(self, mode):
        self.mode = mode

    def answer(self):
        if self.mode == "ready":
            return "Bonjour"
        elif self.mode == "busy":
            raise TelephoneError.Busy()
        else:
            raise ValueError("Testing an unexpected error")

# This is a normal Rust trait - very much like a callback but can be implemented
# in Rust or in foreign code and is generally more consistent with the uniffi
# Arc<>-based object model.
class DiscountSim(SimCard):
    def name(self):
        return "python"

class CallbacksTest(unittest.TestCase):
    TelephoneImpl = Telephone
    def test_answer(self):
        cb_object = CallAnswererImpl("ready")
        telephone = self.TelephoneImpl()
        self.assertEqual("Bonjour", telephone.call(get_sim_cards()[0], cb_object))

    def test_busy(self):
        cb_object = CallAnswererImpl("busy")
        telephone = self.TelephoneImpl()
        with self.assertRaises(TelephoneError.Busy):
            telephone.call(get_sim_cards()[0], cb_object)

    def test_unexpected_error(self):
        cb_object = CallAnswererImpl("something-else")
        telephone = self.TelephoneImpl()
        with self.assertRaises(TelephoneError.InternalTelephoneError):
            telephone.call(get_sim_cards()[0], cb_object)

    def test_sims(self):
        cb_object = CallAnswererImpl("ready")
        telephone = self.TelephoneImpl()
        sim = DiscountSim()
        self.assertEqual("python est bon marché", telephone.call(sim, cb_object))

class FancyCallbacksTest(CallbacksTest):
    TelephoneImpl = FancyTelephone

unittest.main()

File: examples/callbacks/tests/bindings/test_callbacks.rb
# frozen_string_literal: true

# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/. */

require 'test/unit'
require 'callbacks'

# This is defined in UDL as a "callback". It's not possible to have a Rust
# implementation of a callback, they only exist on the foreign side.
class CallAnswererImpl < Callbacks::CallAnswerer
  def initialize(mode)
    @mode = mode
  end

  def answer
    if @mode == 'ready'
      'Bonjour'
    elsif @mode == 'busy'
      raise Callbacks::TelephoneError::Busy
    else
      raise ValueError, 'Testing an unexpected error'
    end
  end
end

# This is a normal Rust trait - very much like a callback but can be implemented
# in Rust or in foreign code and is generally more consistent with the uniffi
# Arc<>-based object model.
class DiscountSim < Callbacks::SimCard
  def name
    'ruby'
  end
end

class TestCallbacks < Test::Unit::TestCase
  TelephoneImpl = Callbacks::Telephone

  def test_answer
    cb_object = CallAnswererImpl.new 'ready'

    assert_equal 'Bonjour', telephone.call(sim, cb_object)
  end

  def test_busy
    cb_object = CallAnswererImpl.new 'busy'

    assert_raise(Callbacks::TelephoneError::Busy) do
      telephone.call sim, cb_object
    end
  end

  def test_unexpected_error
    cb_object = CallAnswererImpl.new 'something-else'

    assert_raise(Callbacks::TelephoneError::InternalTelephoneError) do
      telephone.call sim, cb_object
    end
  end

  def test_sims
    cb_object = CallAnswererImpl.new 'ready'
    sim = DiscountSim.new

    assert_equal 'ruby est bon marché', telephone.call(sim, cb_object)
  end

  def sim
    Callbacks.get_sim_cards[0]
  end

  def telephone
    self.class::TelephoneImpl.new
  end
end

class FancyTestCallbacks < TestCallbacks
  TelephoneImpl = Callbacks::FancyTelephone

  def test_answer
    super
  end

  def test_busy
    super
  end

  def test_unexpected_error
    super
  end

  def test_sims
    super
  end
end
